### PR TITLE
Azure Storage & Queue Plugins

### PR DESCRIPTION
**Pull request by tjholm**
Queue and storage plugins combined, as they share underlying infrastructure (azure storage accounts).

Closes #157

### CODE REVIEW

File: pkg/plugins/storage/azblob/azblob.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+	spt, err := azureutils.GetServicePrincipalToken(azure.PublicCloud.ResourceIdentifiers.Storage)
+	if err != nil {
+		return nil, err
+	}
+
+	cTkn := azblob.NewTokenCredential(spt.Token().AccessToken, tokenRefresherFromSpt(spt))
+
+	var accountURL *url.URL
+	if accountURL, err = url.Parse(fmt.Sprintf("https://%s.blob.core.windows.net", storageAccountName)); err != nil {
```
Comment (raksiv) [inline pkg/plugins/storage/azblob/azblob.go:175]: What if the user has a custom domain ? 
Would it be better to have "https://%s.blob.core.windows.net" exposed as an ENV configuration?

File: pkg/plugins/queue/azqueue/azqueue_test.go [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+	"fmt"
+	//"bytes"
+	//"fmt"
+	"time"
+
+	azqueue2 "github.com/Azure/azure-storage-queue-go/azqueue"
+	"github.com/nitric-dev/membrane/pkg/plugins/queue"
+
+	//"io/ioutil"
+	//"strings"
```
Comment (raksiv) [inline pkg/plugins/queue/azqueue/azqueue_test.go:27]: Minor tidy up to remove unwanted comments.

File: pkg/plugins/storage/azblob/azblob.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+			codes.Internal,
+			"Unable to delete blob",
+			err,
+		)
+	}
+
+	return nil
+}
+
+const expiryBuffer = 2 * time.Minute
```
Comment (raksiv) [inline pkg/plugins/storage/azblob/azblob.go:139]: Assume this doesn't require config? 2 mins is always adequate for the buffer?

**Review by raksiv — CHANGES_REQUESTED** [review]

File: pkg/plugins/queue/azqueue/azqueue.go [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+// Returns an adapted azqueue MessageIdUrl, which is a client for interacting with a specific message (task) in a specific queue
+func (s *AzqueueQueueService) getMessageIdUrl(queue string, messageId azqueue.MessageID) azqueueserviceiface.AzqueueMessageIdUrlIface {
+	mUrl := s.getMessagesUrl(queue)
+
+	return mUrl.NewMessageIDURL(messageId)
+}
+
+func (s *AzqueueQueueService) Send(queue string, task queue.NitricTask) error {
+	newErr := errors.ErrorsWithScope(
+		"AzqueueService.Send",
```
Comment (medgar-nitric) [inline pkg/plugins/queue/azqueue/azqueue.go:61]: "AzqueueQueueService.Send"

File: pkg/plugins/queue/azqueue/azqueue.go [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+
+	return tasks, nil
+}
+
+// Complete - Completes a previously popped queue item
+func (s *AzqueueQueueService) Complete(queue string, leaseId string) error {
+	newErr := errors.ErrorsWithScope(
+		"AzqueueQueueService.Complete",
+		map[string]interface{}{
+			"queue": queue,
```
Comment (medgar-nitric) [inline pkg/plugins/queue/azqueue/azqueue.go:210]: add leaseId parameter

File: pkg/plugins/storage/azblob/azblob.go [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+		} else {
+			tkn := spt.Token()
+			credential.SetToken(tkn.AccessToken)
+
+			return tkn.Expires().Sub(time.Now().Add(expiryBuffer))
+		}
+
+		// Mark the token as already expired
+		return time.Duration(0)
+	}
```
Comment (medgar-nitric) [inline pkg/plugins/storage/azblob/azblob.go:154]: Is there any scope for having some common Azure token refresh code?

**Review by medgar-nitric — CHANGES_REQUESTED** [review]

File: pkg/plugins/storage/azblob/azblob.go (hunk 6 above)
Comment (tjholm) [reply pkg/plugins/storage/azblob/azblob.go:154]: There is but need to examine the interface chain for the structs that implement these tokens for both azqueue and azblob to see where there is commonality, which can be done in a later PR.

File: pkg/plugins/queue/azqueue/azqueue.go (hunk 4 above)
Comment (tjholm) [reply pkg/plugins/queue/azqueue/azqueue.go:61]: done

File: pkg/plugins/queue/azqueue/azqueue.go (hunk 5 above)
Comment (tjholm) [reply pkg/plugins/queue/azqueue/azqueue.go:210]: Could leaseId be considered sensitive? This is something only the holder should know and is not necessarily common across a stack (like queue name).

File: pkg/plugins/queue/azqueue/azqueue_test.go (hunk 2 above)
Comment (tjholm) [reply pkg/plugins/queue/azqueue/azqueue_test.go:27]: done

File: pkg/plugins/storage/azblob/azblob.go (hunk 3 above)
Comment (tjholm) [reply pkg/plugins/storage/azblob/azblob.go:139]: For now, can address if it becomes a problem in future.

File: pkg/plugins/storage/azblob/azblob.go (hunk 1 above)
Comment (tjholm) [reply pkg/plugins/storage/azblob/azblob.go:175]: done, set for both queue and blob, and adding as part of env variables for app service deployment.

File: pkg/plugins/queue/azqueue/azqueue.go (hunk 5 above)
Comment (tjholm) [reply pkg/plugins/queue/azqueue/azqueue.go]: done, might be worth considering if leaseId is sensitive information though as only its holder should know this. (not common stack knowledge like queue names for example.

File: pkg/plugins/queue/azqueue/azqueue.go [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+	mUrl := s.getMessagesUrl(queue)
+
+	return mUrl.NewMessageIDURL(messageId)
+}
+
+func (s *AzqueueQueueService) Send(queue string, task queue.NitricTask) error {
+	newErr := errors.ErrorsWithScope(
+		"AzqueueQueueService.Send",
+		map[string]interface{}{
+			"queue": queue,
```
Comment (medgar-nitric) [inline pkg/plugins/queue/azqueue/azqueue.go:63]: Log task parameter as well.

**Review by medgar-nitric — CHANGES_REQUESTED** [review]

File: pkg/plugins/queue/azqueue/azqueue.go (hunk 7 above)
Comment (tjholm) [reply pkg/plugins/queue/azqueue/azqueue.go:63]: done

**Review by medgar-nitric — APPROVED** [review]

**Review by HomelessDinosaur — APPROVED** [review]

**Review by davemooreuws — APPROVED** [review]

**Pull request merged**